modules/tinymce/src/plugins/table/test/ts/browser/ui/TableCaptionTest.ts

```diff
@@ -0,0 +1,178 @@
+import { assert } from 'chai';
+
+import Editor from 'tinymce/core/api/Editor';
+import { EditorEvent } from 'tinymce/core/api/util/EventDispatcher';
+import { TableModifiedEvent } from 'tinymce/plugins/table/api/Events';
+import Plugin from 'tinymce/plugins/table/Plugin';
+import Theme from 'tinymce/themes/silver/Theme';
+import { assertStructureIsRestoredToDefault, clickOnButton, pClickOnMenuItem, setEditorContentTableAndSelection } from '../../module/test/TableModifiersTestUtils';
+
+describe('browser.tinymce.plugins.table.ui.TableCaptionTest', () => {
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    plugins: 'table',
+    indent: false,
+    toolbar: 'tablecaption',
+    base_url: '/project/tinymce/js/tinymce',
+    setup: (editor: Editor) => {
+      editor.on('tablemodified', logEvent);
+    },
+    menu: {
+      table: { title: 'Table', items: 'tablecaption' },
+    },
+    menubar: 'table',
+  }, [ Plugin, Theme ], true);
+
+  let events: Array<EditorEvent<TableModifiedEvent>> = [];
+  const logEvent = (event: EditorEvent<TableModifiedEvent>) => {
+    events.push(event);
+  };
+
+  afterEach(() => {
+    events = [];
+  });
+
+  const tableCaptionStructureWithCaption = (
+    '<table>' +
+      '<caption>Caption</caption>' +
+      '<tbody>' +
+        '<tr>' +
+          '<td>Filler</td>' +
+        '</tr>' +
+      '</tbody>' +
+    '</table>'
+  );
+
+  const nestedTableCaptionWithCaptionsStructure = (
+    '<table>' +
+      '<caption>Caption</caption>' +
+      '<tbody>' +
+        '<tr>' +
+          '<td>' +
+            '<table>' +
+              '<caption>Caption</caption>' +
+              '<tbody>' +
+                '<tr>' +
+                  '<td>Filler</td>' +
+                '</tr>' +
+              '</tbody>' +
+            '</table>' +
+          '</td>' +
+        '</tr>' +
+      '</tbody>' +
+    '</table>'
+  );
+
+  const nestedTableCaptionStructure = (
+    '<table>' +
+      '<tbody>' +
+        '<tr>' +
+          '<td>' +
+            '<table>' +
+              '<caption>Caption</caption>' +
+              '<tbody>' +
+                '<tr>' +
+                  '<td>Filler</td>' +
+                '</tr>' +
+              '</tbody>' +
+            '</table>' +
+          '</td>' +
+        '</tr>' +
+      '</tbody>' +
+    '</table>'
+  );
+
+  const toggleCaption = (editor: Editor) => {
+    editor.execCommand('mceTableToggleCaption');
+
+    assert.lengthOf(events, 1, 'Command executed successfully');
+  };
+
+  const assertTableStructureIsCorrect = (editor: Editor, tableStructure: string) => {
+    TinyAssertions.assertContent(editor, tableStructure);
+  };
+
+  const setTableCaptionStructureAndSelection = (editor: Editor, structure: string, indexOftbody: number) => {
+    editor.setContent(structure);
+
+    TinySelections.setSelection(editor, [ 0, indexOftbody, 0, 0 ], 0, [ 0, indexOftbody, 0, 0 ], 1);
+  };
+
+  const pAssertTableCaption = async (toolbar: boolean, addCaption: boolean) => {
```

Comment:
Rather than using a boolean for toolbar, could we do something like this

```suggestion
  const pAssertTableCaption = async (action: 'toolbar' | 'menuitem', addCaption: boolean) => {
```

so that it's a bit easier to follow in the tests?

Reply:
Not sure it'll do much good since one of the values will never be seen in the actual code, but sure.